sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala

```diff
@@ -407,7 +410,17 @@ private[sql] class ParquetRelation2(
       val filesToTouch =
         if (shouldMergeSchemas) {
           // Also includes summary files, 'cause there might be empty partition directories.
-          (metadataStatuses ++ commonMetadataStatuses ++ dataStatuses).toSeq
+
+          // If skipMergePartFiles config is true, we assume that all part-files are the same for
+          // their schema with summary files, so we ignore them when merging schema.
+          // If the config is false, which is the default setting, we merge all part-files.
+          val needMerged: Seq[FileStatus] =
+            if (skipMergePartFiles) {
+              Seq()
+            } else {
+              dataStatuses
+            }
+          (metadataStatuses ++ commonMetadataStatuses ++ needMerged).toSeq
```

Comment:
Summary files are not always available. It's quite possible that `metadataStatuses` and `commonMetadataStatuses` are both empty.